File: sort/c/selection_sort.c
/* 
 * selection_sort.c
 * Perform insertion sort on an array of integers
 */

#include "sort.h"

static void swap(int *a, int *b) {
    int temp;
    temp = *a;
    *a = *b; 
    *b = temp;
}

void selection_sort(int arr[], int len) {
    int i, current_min, current_min_index, j;
    
    for(i = 0; i < len; i++) {
        current_min_index = i; 
        current_min = arr[current_min_index];
        for(j = i + 1; j < len; j++) {
            if(arr[j] < current_min) {
                current_min_index = j; 
                current_min = arr[j];
            }
        }
        if(current_min_index != i) {
            swap(&arr[i], &arr[current_min_index]);
        } 
    }
} 

File: sort/java/SelectionSort.java
/** 
 * This class implements SelectionSort on an array of integers. 
 * Selection sort is a simple sorting algorithm, which has an O(n^2) 
 * time complexity
 */

public class SelectionSort {

    private static void swapArrayElements(int [] arr, int index1, int index2) {
        if(index1 > arr.length || index1 < 0 || index2 > arr.length || index2 < 0) {
            throw new IndexOutOfBoundsException(); 	
        }
        int temp = arr[index1];
        arr[index1] = arr[index2];
        arr[index2] = temp;
    }

    public static void sort(int [] arr) {
        int currentMinimum, currentMinimumIndex;
        for(int i = 0; i < arr.length; i++) {
            currentMinimumIndex = i;
            currentMinimum = arr[i];

            for(int j = i + 1; j < arr.length; j++) {
                if(arr[j] < currentMinimum) {
                    currentMinimumIndex = j;
                    currentMinimum = arr[j];
                }
            }

            if(i != currentMinimumIndex) {
                swapArrayElements(arr, i, currentMinimumIndex);			
            }
        }
    }
} 
